cli/src/action/database/upgrade/mod.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright 2018-2021 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Provides database upgrade functionality
+
+#[cfg(feature = "node-id-upgrade")]
+mod node_id;
+mod yaml;
+
+use std::path::PathBuf;
+use std::str::FromStr;
+
+use clap::ArgMatches;
+
+#[cfg(feature = "sqlite")]
+use crate::action::database::sqlite::{get_database_at_state_path, get_default_database};
+use crate::action::database::{ConnectionUri, PgConnection, SplinterEnvironment};
+use crate::diesel::Connection;
+use crate::error::CliError;
+
+use super::Action;
+
+/// The overarching Action possibly containing multiple upgrade actions
+pub struct UpgradeAction;
+
+impl Action for UpgradeAction {
+    fn run<'a>(&mut self, arg_matches: Option<&ArgMatches<'a>>) -> Result<(), CliError> {
+        let state_dir = get_state_dir(arg_matches)?;
+        let database_uri = get_database_uri(arg_matches)?;
+        let connection_uri = database_uri.parse().map_err(|_| {
```

Comment:
Why not return a `ConnectionUri` from the `get_database_uri` function - it's already parsed as such to check connectivity.

Reply:
Updated to return a ConnectionUri